### PR TITLE
explain: print target cluster in `AS TEXT` plan output

### LINKED ISSUE #24641: Include target cluster in `EXPLAIN <statement>` output
### Feature request

It would be helpful if `EXPLAIN <statement>` can print the determined target cluster for a statement:

```text
<plan ...>

Target cluster: quickstart
```

This can be done behind an optional `WITH(target cluster)` config flag at first.

### PR DESCRIPTION
**Pull request by aalexandrov**
With the #25834 currently in progress, I thought it was a good time to add support for `EXPLAIN WITH(target cluster)` so we can test / verify that the new `mz_~` clusters are executing the queries as expected (for example @jubrad might find this useful).

### Motivation

  * This PR adds a known-desirable feature.

Closes #24641.

### Tips for reviewer

- @jkosh44: This is mostly trivial plumbing.
- cc @morsapaes: I've also used the occasion to update the docs, which should also close a TODO in #25273.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Add an `EXPLAIN WITH (target cluster)` parameter in order to show the cluster selected for a query.

### CODE REVIEW

**Comment by aalexandrov** [comment]
Exercising the feature on my locally running `environmentd` reveals that a selection from an unindexed user table goes to `mz_introspection`, `mz_system`, or `quickstart` depending on the user. I'm not sure whether the discrepancy between the `mz_support` and `mz_system` connections is expected.

```text
materialize=> explain with(target cluster) select current_user(), * from t1;
             Optimized Plan              
-----------------------------------------
 Explained Query:                       +
   Project (#2, #0, #1)                 +
     Map ("mz_support")                 +
       ReadStorage materialize.public.t1+
                                        +
 Target cluster: mz_introspection       +
 
(1 row)


materialize=> explain with(target cluster) select current_user(), * from t1;
             Optimized Plan              
-----------------------------------------
 Explained Query:                       +
   Project (#2, #0, #1)                 +
     Map ("mz_system")                  +
       ReadStorage materialize.public.t1+
                                        +
 Target cluster: mz_system              +
 
(1 row)


materialize=> explain with(target cluster) select current_user(), * from t1;
             Optimized Plan              
-----------------------------------------
 Explained Query:                       +
   Project (#2, #0, #1)                 +
     Map ("materialize")                +
       ReadStorage materialize.public.t1+
                                        +
 Target cluster: quickstart             +
 
(1 row)
```

File: src/sql-parser/src/ast/defs/statement.rs [hunk 1]
```diff
@@ -3217,6 +3217,7 @@ pub enum ExplainPlanOptionName {
     Raw, // Listed after the `Raw~` variants to keep the parser happy!
     Redacted,
     SubtreeSize,
+    TargetCluster,
```
Comment (jkosh44) [inline src/sql-parser/src/ast/defs/statement.rs:3220]: Can you add a parser test for this?

**Comment by jkosh44** [comment]
Does this work with transaction? For example if I ran the following:

```SQL
BEGIN;
SELECT * FROM t;
EXPLAIN (TARGET CLUSTER) FOR SELECT * FROM mz_views;
```

Would it correctly identify the target cluster is not auto-routed?

**Comment by jkosh44** [comment]
> Exercising the feature on my locally running `environmentd` reveals that a selection from an unindexed user table `t1` goes to `mz_introspection`, `mz_system`, or `quickstart` depending on the user. I'm not sure whether the discrepancy between the `mz_support` and `mz_system` sessions is expected.

It is expected. The `mz_system` role has it's `cluster` default session variable overridden to `mz_system` and the `mz_support` role has it's `cluster` default session variable overridden to `mz_introspection`.

File: src/sql-parser/src/ast/defs/statement.rs (hunk 1 above)
Comment (jkosh44) [reply src/sql-parser/src/ast/defs/statement.rs:3220]: Up to you, it's not a big deal either way.

**Review by jkosh44 — APPROVED** [review]
LGTM

**Comment by jkosh44** [comment]
> > Would it correctly identify the target cluster is not auto-routed?
> 
> Yes, it returns the `quickstart` cluster (I guess this is forced by the execution of the `SELECT * FROM t` query above).

Yeah, once a cluster is picked in a transaction then it cannot change. So the first query picks the `quickstart` cluster and all later queries in that transaction must also use the `quickstart` cluster. It's probably worth adding a test like this if you haven't already.

**Review by benesch — COMMENTED** [review]
Why do we want to put this behind an EXPLAIN option? This seems useful enough to have in the explain output by default, and there doesn’t seem to be any downside?

**Comment by def-** [comment]
+1 on putting this in any explain output, then I don't have to add testing for this but we get all the existing testing for free.

**Review by def- — APPROVED** [review]

**Comment by aalexandrov** [comment]
@MaterializeInc/testing if the Nightlies continue to fail, I would really appreciate if I can get some help in setting the right `if` predicates in order to get the `remove_target_cluster_from_explain(...)` calls right.

It took me ~1/2 days to implement the feature behind an explain flag, but since then I've spent the better part of two days trying to fix all CI tests after removing the explain flag.

**Comment by aalexandrov** [comment]
There is really no good way of for example fixing the condition in the `HydrateIndex` benchmark scenario in a way that will allow me to test correctly with a Nightlies run against this PR. For example I now have

```python
if self._mz_version < MzVersion.parse_mz("v0.96.0-dev"):
    sql = remove_target_cluster_from_explain(sql)
```

which should remove the

```
Target cluster: ....
```

lines from the Testdrive snippet for older materialize versions. However when I run Nightlies in CI with

```bash
bin/mzcompose --find feature-benchmark run default --other-tag common-ancestor
```

the common-ancestor gets resolved to another commit that has the `v0.96.0` version.

**Comment by aalexandrov** [comment]
Setting this to auto-merge because I cannot differentiate which of my tests fail because I didn't fix them correctly and which will because the "other" version for the PR is the common ancestor. Will fix what is broken in Nightlies in the next days.

**Pull request merged**